### PR TITLE
Added stochastic depth layer

### LINKED ISSUE #2032: Feature Request : Stochastic Depth/ResDrop
**Describe the feature and the current behavior/state.**
Stochastic depth is a regularization technique to train very deep residual networks (the authors for example train ResNets of up to 1202 layers). Particularly, it allows training of, on average, more shallow networks, that retain their full depth at inference time.

Previous attempts (https://github.com/tensorflow/addons/issues/626) seem to have been unnecessarily convoluted. This implemenation would be a simple layer, attached to the end of a residual branch (as suggested here https://github.com/tensorflow/tensorflow/issues/8817#issuecomment-290788333, which is also the way it is described in the Shake-Drop paper: https://arxiv.org/abs/1802.02375)

Noteably this means that, if momentum is being used, the dropped layers will still receive small updates due to their historic gradients (similar to Dropout).

**Relevant information**
- Are you willing to contribute it (yes/no): Yes
- Are you willing to maintain it going forward? (yes/no): Yes
- Is there a relevant academic paper? (if so, where): Yes (https://arxiv.org/abs/1603.09382)
- Is there already an implementation in another framework? (if so, where): The authors original pytorch implementation is available here: https://github.com/yueatsprograms/Stochastic_Depth
- Was it part of tf.contrib? (if so, where): No

**Which API type would this fall under (layer, metric, optimizer, etc.)**
layer
**Who will benefit with this feature?**
Anybody who wants to use stochastic depth to train deeper ResNets, or who wants to recreate the EfficientNet architecture (https://arxiv.org/abs/1905.11946). Anyone who wants to add Shake-Drop (https://arxiv.org/abs/1802.02375) to their network, can use this as base.
**Any other info.**
It is important to note that Stochastic Depth is not the same as Dropout with noise_shape=(1, 1, 1), as suggested in the tensorflow EfficientNet implementation (https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/applications/efficientnet.py#L511). 
Using dropout, the branch is simply kept as is during inference time, however for stochastic depth, the branch is re-scaled based on its survival probability, before being merged with the main network.

### PR DESCRIPTION
**Pull request by MHStadler**
# Description

Brief Description of the PR:

Resolves #2032

## Type of change

- [ ] Bug fix
- [ ] New Tutorial
- [ ] Updated or additional documentation
- [ ] Additional Testing
- [ ] New Activation and the changes conform to the [activation contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/activations/README.md#contribution-guidelines)
- [ ] New Callback and the changes conform to the [callback contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/callbacks/README.md#contribution-guidelines)
- [ ] New Image addition and the changes conform to the [image op contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/image/README.md#contribution-guidelines)
- [x] New Layer and the changes conform to the [layer contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/layers/README.md#contribution-guidelines)
- [ ] New Loss and the changes conform to the [loss contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/losses/README.md#contribution-guidelines)
- [ ] New Metric and the changes conform to the [metric contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/metrics/README.md#contribution-guidelines)
- [ ] New Optimizer and the changes conform to the [optimizer contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/optimizers/README.md#contribution-guidelines)
- [ ] New RNN Cell and the changes conform to the [rnn contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/rnn/README.md#contribution-guidelines)
- [ ] New Seq2seq addition and the changes conform to the [seq2seq contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/seq2seq/README.md#contribution-guidelines)
- [ ] New Text addition and the changes conform to the [text op contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/text/README.md#contribution-guidelines)

# Checklist:

- [x] I've properly [formatted my code according to the guidelines](https://github.com/tensorflow/addons/blob/master/CONTRIBUTING.md#coding-style)
    - [ ] By running Black + Flake8
    - [x] By running pre-commit hooks
- [x] This PR addresses an already submitted issue for TensorFlow Addons
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] This PR contains modifications to C++ custom-ops

# How Has This Been Tested?

* Unit Tests verifying the functionality 
* Unit Test verifying serialization and deserialization

### CODE REVIEW

**Comment by bhack** [comment]
We need a CODEOWNER for this and a co-CODEOWNER as sponsor.

**Comment by WindQAQ** [comment]
I can be the co-owner, so @MHStadler feels free to list you (if you wish) and me in the `.github/CODEOWNERS` :-)

**Comment by MHStadler** [comment]
@WindQAQ Thanks, I added us two

File: tensorflow_addons/layers/stochastic_depth.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    return tf.keras.layers.Add()([input, residual])
+    ```
+
+    StochasticDepth acts as a drop-in replacement for the addition:
+
+    ```python
+    residual = tf.keras.layers.Conv2D(...)(input)
+
+    return tfa.layers.StochasticDepth()([input, residual])
+    ```
```
Comment (WindQAQ) [inline tensorflow_addons/layers/stochastic_depth.py:29]: For python code snippet, we encourage to use testable docstring with doctest. See https://github.com/tensorflow/addons/blob/master/CONTRIBUTING.md#testing-docstrings .

File: tensorflow_addons/layers/stochastic_depth.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    x[0] + b_l * x[1]
+    ```
+
+    , where b_l is a random Bernoulli variable with probability p(b_l == 1) == p_l
+
+    At test time, StochasticDepth rescales the activations of the residual branch based on the survival probability:
+
+    ```python
+    x[0] + p_l * x[1]
+    ```
```
Comment (WindQAQ) [inline tensorflow_addons/layers/stochastic_depth.py:43]: I guess you want to use mathjax expression here and above, you can wrap your math expr with `$\alpha$` for inline math expr or
```
$$
\alpha
$$
```
for blocked math expr.

File: tensorflow_addons/layers/stochastic_depth.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    """
+
+    @typechecked
+    def __init__(self, p_l: float = 0.5, **kwargs):
+        super().__init__(**kwargs)
+
+        self.p_l = p_l
+
+    def call(self, x, training=None):
+        assert isinstance(x, list)
```
Comment (WindQAQ) [inline tensorflow_addons/layers/stochastic_depth.py:63]: Maybe using
```
if not isinstance(x, list):
    raise ValueError("xxx")
```
can give more informative error.
Also, you might want to check the length of this list too.

File: tensorflow_addons/layers/tests/stochastic_depth_test.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
+import pytest
+
+import numpy as np
+import tensorflow as tf
+from tensorflow_addons.layers.stochastic_depth import StochasticDepth
+from tensorflow_addons.utils import test_utils
+
+KEEP_SEED = 1111
+DROP_SEED = 2222
```
Comment (WindQAQ) [inline tensorflow_addons/layers/tests/stochastic_depth_test.py:9]: Prefix this with underscore.
```
_KEEP_SEED = 1111
_DROP_SEED = 2222
```

File: tensorflow_addons/layers/stochastic_depth.py [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+    Output shape:
+        Equal to the shape of inputs `shortcut`, and `residual`
+    """
+
+    @typechecked
+    def __init__(self, p_l: float = 0.5, **kwargs):
+        super().__init__(**kwargs)
+
+        self.p_l = p_l
```
Comment (WindQAQ) [inline tensorflow_addons/layers/stochastic_depth.py:60]: Though it's the name from the paper, I think we can have more clear naming for this argument.

**Comment by bhack** [comment]
Do we need mixed precision tests?

File: tensorflow_addons/layers/stochastic_depth.py [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    [Deep Networks with Stochastic Depth](https://arxiv.org/abs/1603.09382), to randomly drop residual branches
+    in residual architectures.
+
+    Usage:
+    Residual architectures with fixed depth, use residual branches that are merged back into the main network
+    by adding the residual branch back to the input:
+
+    >>> input = np.ones((1, 3, 3, 1), dtype = np.float32)
+    >>> residual = tf.keras.layers.Conv2D(1, 1)(input)
+    >>> output = tf.keras.layers.Add()([input, residual])
```
Comment (WindQAQ) [inline tensorflow_addons/layers/stochastic_depth.py:19]: Can we print output.shape, that is,
```
>>> output.shape
```

File: tensorflow_addons/layers/stochastic_depth.py [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+
+    StochasticDepth acts as a drop-in replacement for the addition:
+
+    >>> input = np.ones((1, 3, 3, 1), dtype = np.float32)
+    >>> residual = tf.keras.layers.Conv2D(1, 1)(input)
+    >>> output = tfa.layers.StochasticDepth()([input, residual])
```
Comment (WindQAQ) [inline tensorflow_addons/layers/stochastic_depth.py:25]: ditto

File: tensorflow_addons/layers/stochastic_depth.py [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+
+    At train time, StochasticDepth returns:
+
+    $$
+    x[0] + b_l * x[1]
+    $$
+
+    , where $b_l$ is a random Bernoulli variable with probability $p(b_l == 1) == p_l$
```
Comment (WindQAQ) [inline tensorflow_addons/layers/stochastic_depth.py:33]: NIT: $P(b_l = 1) = p_l$

File: tensorflow_addons/layers/stochastic_depth.py [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+
+    At test time, StochasticDepth rescales the activations of the residual branch based on the survival probability ($p_l$):
+
+    $$
+    x[0] + p_l * x[1]
+    $$
+
+    Arguments:
+        survival_probability: float, the probability of the residual branch being kept.
```
Comment (WindQAQ) [inline tensorflow_addons/layers/stochastic_depth.py:42]: Maybe we can use `rate`? `tf.keras.layers.{Dropout,GaussianDropout}` both use this term.

File: tensorflow_addons/layers/stochastic_depth.py [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+
+    Call Arguments:
+        inputs:  List of `[shortcut, residual]` where
+            * `shortcut`, and `residual` are tensors of equal shape.
```
Comment (WindQAQ) [inline tensorflow_addons/layers/stochastic_depth.py:46]: `*` is redundant.

File: tensorflow_addons/layers/stochastic_depth.py [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+
+    @typechecked
+    def __init__(self, survival_probability: float = 0.5, **kwargs):
+        super().__init__(**kwargs)
+
+        self.survival_probability = survival_probability
+
+    def call(self, x, training=None):
+        assert isinstance(x, list), "Input must be a list"
+        assert len(x) == 2, "Input must have exactly two entries"
```
Comment (WindQAQ) [inline tensorflow_addons/layers/stochastic_depth.py:60]:
```python
if not is instance(x, list) or len(x) != 2:
    raise ValueError("input must be a list of length 2.")
```

File: tensorflow_addons/layers/tests/stochastic_depth_test.py [hunk 12]
```diff
@@ -0,0 +1,58 @@
+import pytest
+
+import numpy as np
+import tensorflow as tf
+from tensorflow_addons.layers.stochastic_depth import StochasticDepth
+from tensorflow_addons.utils import test_utils
```
Comment (WindQAQ) [inline tensorflow_addons/layers/tests/stochastic_depth_test.py:6]:
```python
import pytest
import numpy as np
import tensorflow as tf

from tensorflow_addons.layers.stochastic_depth import StochasticDepth
from tensorflow_addons.utils import test_utils

File: tensorflow_addons/layers/stochastic_depth.py (hunk 9 above)
Comment (MHStadler) [reply tensorflow_addons/layers/stochastic_depth.py:42]: Hm, the problem is that Dropout and GaussianDropout denote the "drop" rate, but it's the opposite here (it's not the chance of dropping, it's the chance of keeping it). Alternatively, I could do rate same as dropout, and then use 1 - p_l in the random variable?

Comment (WindQAQ) [reply tensorflow_addons/layers/stochastic_depth.py:42]: Agree

Comment (WindQAQ) [reply tensorflow_addons/layers/stochastic_depth.py:42]: oops sorry, I see the paper use the term `survival_probability`, let's keep it the same.

**Comment by MHStadler** [comment]
@bhack I looked at some of the existing mixed_precision tests, and I added one to check the output dtype is retained correctly 

Let me know if and which further tests I should add

File: tensorflow_addons/layers/stochastic_depth.py [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+    >>> output.shape
+    TensorShape([1, 3, 3, 1])
+
+    At train time, StochasticDepth returns:
+
+    $$
+    x[0] + b_l * x[1]
+    $$
+
+    , where $b_l$ is a random Bernoulli variable with probability $P(b_l = 1) = p_l$
```
Comment (WindQAQ) [inline tensorflow_addons/layers/stochastic_depth.py:37]: NIT: move `,` into the math block.
$$
x[0] + b_l * x[1],
$$

File: tensorflow_addons/layers/stochastic_depth.py [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+        def _call_test():
+            return shortcut + self.survival_probability * residual
+
+        return tf.keras.backend.in_train_phase(
+            _call_train, _call_test, training=training
+        )
+
+    def compute_output_shape(self, input_shape):
+        if not isinstance(input_shape, list) or len(input_shape) != 2:
+            raise ValueError("input_shape must be a list of length 2.")
```
Comment (WindQAQ) [inline tensorflow_addons/layers/stochastic_depth.py:82]: I think this check is somehow redundant because we already have it in `call`.

File: tensorflow_addons/layers/stochastic_depth.py [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+        self.survival_probability = survival_probability
+
+    def call(self, x, training=None):
+        if not isinstance(x, list) or len(x) != 2:
+            raise ValueError("input must be a list of length 2.")
+
+        shortcut, residual = x
+
+        # Random bernoulli variable indicating whether the branch should be kept or not or not
+        b_l = tf.keras.backend.random_binomial([], p=self.survival_probability)
```
Comment (WindQAQ) [inline tensorflow_addons/layers/stochastic_depth.py:68]: This is going to deprecate https://www.tensorflow.org/api_docs/python/tf/keras/backend/random_binomial. We should change to `tf.keras.backend.random_bernoulli`.

**Review by WindQAQ — APPROVED** [review]
Thanks for the contribution!
